### PR TITLE
[SPARK-43937][CONNECT][PYTHON] Add ifnull,isnotnull,equal_null,nullif,nvl,nvl2 to Scala and Python

### PR DESCRIPTION
**Pull request by panbingkun**
### What changes were proposed in this pull request?
Add following functions:

- ~~not: already exists on the Scala side, but is it still not on the Python side due to keyword conflicts?~~
- ~~if: new add on the scala side, however it conflicts with Python keywords `if` and cannot be added on the Python side?~~
- ifnull
- isnotnull
- equal_null
- nullif
- nvl
- nvl2

to:

- Scala API
- Python API
- Spark Connect Scala Client
- Spark Connect Python Client

### Why are the changes needed?
for parity

### Does this PR introduce _any_ user-facing change?
Yes, new functions.

### How was this patch tested?
- Add New UT.

### CODE REVIEW

**Comment by panbingkun** [comment]
cc @zhengruifeng

File: python/pyspark/sql/functions.py [hunk 1]
```diff
@@ -11157,6 +11157,142 @@ def udf(
         return _create_py_udf(f=f, returnType=returnType, useArrow=useArrow)
 
 
+# ---------------  Conditional functions ------------------------
+
```
Comment (panbingkun) [inline python/pyspark/sql/functions.py:11161]: `if`: new add on the scala side, however it conflicts with Python keywords `if` and cannot be added on the Python side?

Comment (zhengruifeng) [reply python/pyspark/sql/functions.py:11161]: I think we should apply a different name in this case, like `map` in Scala <-> `create_map` in Python.

cc @HyukjinKwon @beliefer do you have a good name for it?

File: python/pyspark/sql/functions.py [hunk 2]
```diff
@@ -11157,6 +11157,136 @@ def udf(
...
+
+@try_remote_functions
+def ifnull(col1: "ColumnOrName", col2: "ColumnOrName") -> Column:
```
Comment (zhengruifeng) [inline python/pyspark/sql/functions.py:11164]: current `ifnull` use the same expression as `nvl`, let's discuss in https://github.com/apache/spark/pull/41516#discussion_r1225030166 first

File: connector/connect/client/jvm/src/main/scala/org/apache/spark/sql/functions.scala [hunk 3]
```diff
@@ -5883,4 +5883,65 @@ object functions {
...
+  def isnotnull(col: Column): Column = Column.fn("isnotnull", col)
+
+  /**
+   * Returns same result as the EQUAL(=) operator for non-null operands, but returns true if both
+   * are null, false if one of the them is null.
+   *
+   * @group conditional_funcs
+   * @since 3.5.0
+   */
+  def equal_null(col1: Column, col2: Column): Column = Column.fn("equal_null", col1, col2)
```
Comment (beliefer) [inline connector/connect/client/jvm/src/main/scala/org/apache/spark/sql/functions.scala:5922]: I don't know the reason put the `EqualNull` into the group `misc_funcs`. It seems EqualNull should putted into group `predicate_funcs`. cc @cloud-fan @zhengruifeng

File: connector/connect/client/jvm/src/main/scala/org/apache/spark/sql/functions.scala [hunk 4]
```diff
@@ -5883,4 +5883,65 @@ object functions {
...
+   */
+  def ifnull(col1: Column, col2: Column): Column = Column.fn("ifnull", col1, col2)
+
+  /**
+   * Returns true if `col` is not null, or false otherwise.
+   *
+   * @group conditional_funcs
+   * @since 3.5.0
+   */
+  def isnotnull(col: Column): Column = Column.fn("isnotnull", col)
```
Comment (beliefer) [inline connector/connect/client/jvm/src/main/scala/org/apache/spark/sql/functions.scala:5913]: `@group predicate_funcs`

File: connector/connect/client/jvm/src/main/scala/org/apache/spark/sql/functions.scala (hunk 3 above)
Comment (zhengruifeng) [reply connector/connect/client/jvm/src/main/scala/org/apache/spark/sql/functions.scala]: I see a few functions in `FunctionRegistry` (and other places) were placed in wrong group. e.g. `Abs` should be in `math` instead of `misc non-aggregate functions`

File: python/pyspark/sql/functions.py (hunk 2 above)
Comment (zhengruifeng) [reply python/pyspark/sql/functions.py]: let's keep current implementation which directly invoke alias `ifnull`

File: python/pyspark/sql/functions.py (hunk 1 above)
Comment (zhengruifeng) [reply python/pyspark/sql/functions.py]: friendly ping @HyukjinKwon  for the function names for `any` and `if` ?

File: connector/connect/client/jvm/src/main/scala/org/apache/spark/sql/functions.scala (hunk 3 above)
Comment (beliefer) [reply connector/connect/client/jvm/src/main/scala/org/apache/spark/sql/functions.scala]: I think we should fix it together.

File: python/pyspark/sql/functions.py (hunk 1 above)
Comment (zhengruifeng) [reply python/pyspark/sql/functions.py]: @beliefer @panbingkun sorry, but I'd suggest we exclude `any` `if` for now: we have `some/bool_or` to replace `any`, and `when/otherwise` to replace `if`.

let's add them in separate PRs later if we really need them

File: connector/connect/client/jvm/src/main/scala/org/apache/spark/sql/functions.scala [hunk 5]
```diff
@@ -2961,6 +2961,64 @@ object functions {
       allowDifferentLgConfigK: Boolean): Column =
     hll_union(Column(columnName1), Column(columnName2), allowDifferentLgConfigK)
 
+  /**
+   * If `predicate` evaluates to true, then returns `trueValue`; otherwise returns `falseValue`.
+   *
+   * @group misc_funcs
+   * @since 3.5.0
+   */
+  def `if`(predicate: Column, trueValue: Column, falseValue: Column): Column =
```
Comment (zhengruifeng) [inline connector/connect/client/jvm/src/main/scala/org/apache/spark/sql/functions.scala:2970]: let's also exclude `if` on the scala side.

Comment (panbingkun) [reply connector/connect/client/jvm/src/main/scala/org/apache/spark/sql/functions.scala:2970]: ok

File: connector/connect/client/jvm/src/main/scala/org/apache/spark/sql/functions.scala (hunk 3 above)
Comment (panbingkun) [reply connector/connect/client/jvm/src/main/scala/org/apache/spark/sql/functions.scala]: So, let's put these functions in 'predicate_funcs' first? and later I will propose a new PR to fix grouping and other trivial issues. @zhengruifeng @beliefer

Comment (zhengruifeng) [reply connector/connect/client/jvm/src/main/scala/org/apache/spark/sql/functions.scala]: I think it is fine, let's fix the grouping in separate PR

File: python/pyspark/sql/functions.py (hunk 1 above)
Comment (panbingkun) [reply python/pyspark/sql/functions.py]: ok, let me remove it.

File: python/pyspark/sql/tests/test_functions.py [hunk 6]
```diff
@@ -60,7 +60,7 @@ def test_function_parity(self):
             "typedLit",  # Scala only
             "monotonicallyIncreasingId",  # depreciated, use monotonically_increasing_id
             "not",  # equivalent to python ~expression
-            "udaf",  # used for creating UDAF's which are not supported in PySpark
+            "udaf", # used for creating UDAF's which are not supported in PySpark
```
Comment (zhengruifeng) [inline python/pyspark/sql/tests/test_functions.py:63]: let's don't touch unrelated files

Comment (panbingkun) [reply python/pyspark/sql/tests/test_functions.py:63]: Okay, I see an extra space in front, and I can't help but want to delete it, haha
Let me revert it.

Comment (panbingkun) [reply python/pyspark/sql/tests/test_functions.py:63]: Ok, Let me revert it.

File: sql/core/src/main/scala/org/apache/spark/sql/functions.scala [hunk 7]
```diff
@@ -6229,6 +6229,71 @@ object functions {
...
+  // Predicates functions
+  //////////////////////////////////////////////////////////////////////////////////////////////
+
+  /**
+   * Returns `col2` if `col1` is null, or `col1` otherwise.
+   *
+   * @group predicates_funcs
+   * @since 3.5.0
+   */
+  def ifnull(col1: Column, col2: Column): Column = withExpr {
```
Comment (zhengruifeng) [inline sql/core/src/main/scala/org/apache/spark/sql/functions.scala:6242]: the `replacement` in `ifnull`/`nvl`/`nvl2`/`nullif` is missing?

Comment (panbingkun) [reply sql/core/src/main/scala/org/apache/spark/sql/functions.scala:6242]: According to my understanding, 'replacement: Expression' is only an internal Spark mechanism that exists to cooperate with 'RuntimeReplaceable' and is designed to reuse expressions. I guess it is actually not needed in our function.

Comment (beliefer) [reply sql/core/src/main/scala/org/apache/spark/sql/functions.scala:6242]: @zhengruifeng the default constructor with `replacement: Expression` will be replaced by rules. We only need consider the other constructor without `replacement`.

Comment (zhengruifeng) [reply sql/core/src/main/scala/org/apache/spark/sql/functions.scala:6242]: oh, thanks for the explanation! @panbingkun @beliefer

**Review by zhengruifeng — APPROVED** [review]

**Comment by zhengruifeng** [comment]
merged to mater, thank you @panbingkun @beliefer
